keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1149 @@
+and a 1.8x speedup by disabling map key sorting. According to the spec, “the
+CBOR data model for maps does not allow ascribing semantics to the order of the
+key/value pairs in the map representation.”  And since the CBOR decoder will
+reject maps containing duplicate keys, a CBOR map represents exactly the same
+set of key-value pairs regardless of the order they are encoded.
+
+In order to take advantage of the available speedup, the CBOR encoder will
+support separate deterministic and nondeterministic modes. The deterministic
+mode will be used for storage serialization only. The nondeterministic mode
+should actively randomize the order of map item encoding (as with map iteration
+in Go) to prevent users from making assumptions about the order.
+
+Clients may imagine other use cases for deterministic encoding of response
+bodies:
+
+- caching
+  - The existing WithCacheControl filter sets the response header
+  “Cache-Control: no-cache, private” to prevent shared caches from storing
+  responses (since requests are subject to authn/authz), and to prevent
+  responses in non-shared caches from being reused without
+  validation. Deterministic encoding could allow an API server to generate
+  strong ETags by hashing the encoded form of the resource.
+  - Even for the existing data formats, there should be no caching proxies
+  storing API responses.
+- diffing
+  - The human-readable text formats (JSON and YAML) are not changing under this
+  proposal.
+  - The documentation will explicitly state that this should not be done with
+  CBOR.
+
+### Unicode
+
+CBOR supports distinct major types for text strings and byte strings. Text
+strings that do not contain a valid UTF-8 sequence are well-formed but invalid
+CBOR. Unlike JSON strings, CBOR text strings do not support any escape
+sequences.
+
+The JSON serializer replaces invalid UTF-8 sequences with the Unicode
+replacement character (u+fffd) during both encode and decode. This is consistent
+with the behavior of encoding/json in the Go standard library. Generated
+Protobuf marshal and unmarshal code neither validates nor coerces strings; the
+byte sequence is directly copied on both encode and decode.
+
+<<[UNRESOLVED]>>
+To avoid accepting invalid CBOR, the decoder will produce an error if a text
+string is not a valid UTF-8 sequence. There are several options for the encode
+behavior:
+
+1. Encode Go strings to CBOR's byte string type, which can store arbitrary byte
+   sequences. There's precedent for doing this with Protobuf and it's the
+   simplest choice.
+1. Implement UTF-8 coercion on encode/decode similar to what is done for JSON.
+1. Validate string contents at encode time. This isn't practical as certain
+   objects that encode successfully using other serializers will fail to encode
+   using CBOR.
+<<[/UNRESOLVED]>>
+
+### Libraries
+
+|                                      | github.com/ugorji/go/codec | github.com/fxamacker/cbor/v2 |
```

Comment:
As I mentioned in the SIG meeting, I'd like to do some sanity checking so that we know if we will be able to support Python and Java clients.

Reply:
I added roundtripping through non-Go implementations to the list of things to fuzz test. I don't know how much effort might be required for that to consistently pass, but it seems useful as a signal even if it doesn't.